### ISSUE TITLE
atools/fs/pln/flightplanio.cpp has export of GPX with ISO8601 tracklog timestamps in whole seconds - needs milliseconds

### ISSUE THREAD

**Issue by ijl20** [albar965/littlenavmap#907]
Thanks for your awesome freeware.

This note could be considered a feature improvement rather than a bug. Essentially the MSFS 2020 recorded tracklog as exported by LNM in GPX format flips between a 3-second and 2-second period between tracklog entries, although the underlying 'correct' time periods are consistently around 3 seconds. This cause velocity calculations (distance/time) from the tracklog to have considerable jitter. It seems likely the cause is truncating (or rounding) the timestamp to the nearest integer second.

The 'GPX file export'  provides a valid Garmin GPX format output file with correct-format `trkpt` entries inclucing ISO8601 timestamps, e.g.
```
<trkpt lon="7.833391" lat="46.654652">
        <ele>1833.63</ele>
        <time>2022-07-31T06:48:31Z</time>
</trkpt>
```
However the exact-second truncation of the actual flight timestamps causes substantial 'jitter' in any velocity calculation from that tracklog, as the period between entries in the tracklog appears to jump between 2 seconds and 3 seconds.

The ISO8601 datetime format allows decimals in the seconds, e.g. the same entry could instead include:
```
        <time>2022-07-31T06:48:31.123Z</time>
```

QDateTime does support outputting this format (https://doc.qt.io/qt-6/qdatetime.html#toString) and the GPX format explictly allows the 'decimals' format to be used i.e.

atools/fs/pln/flightplanio.cpp line 3185
```
                                  toString("yyyy-MM-ddTHH:mm:ssZ"));
```
could be changed to:
```
                                  toString("yyyy-MM-ddTHH:mm:ss.zzzZ"));
```

I see the same jitter in the 'velocity' line on the altitude/velocity chart displayed in LNM, so it's possible you are converting the timestamps from the flightsim to integers before using them in LNM at all, so the 'fix' suggested by me would only be a part of the puzzle.

**Comment by albar965** [comment]
Seems that GPX the GPX spec allows the milliseconds too: https://www.topografix.com/GPX/1/1/gpx.xsd

Makes sense and allows me to weed out a few 32-bit timestamps in LNM for the next 2.7 alpha (unlikely to be added to the stable 2.6 but I have to check).

**Comment by albar965** [comment]
I just looked into this, and currently no simulator supports accuracy higher than full seconds. :frowning_face:

**Comment by albar965** [comment]
Correction: X-Plane provides sub-second accuracy.

**Comment by ijl20** [comment]
My bad - I think I opened a can of worms thinking it would be a one-liner fix. FSX/MSFS provides high-resolution decimal seconds for the time e.g. ABSOLUTE TIME (seconds), ZULU TIME (seconds), SIMULATOR TIME (seconds MSFS only I think), LOCAL TIME (seconds) so the real-time mods (like mine) can work - (see https://docs.flightsimulator.com/html/Additional_Information/Reverse_Polish_Notation.htm). TBH the most complicated timing piece I've needed to handle in MSFS is when the player hits 'pause', as the time-of-day clock in MSFS keeps running - there's a new MSFS simvar SIMULATION TIME which pauses during a pause but that immediately moves you out of the convenience of an actual Zulu time simvar.
Apologies - my bad - I mistakenly thought your 'internal' times would be decimal and the issue was with the ISO8601 output so suggested the one-liner.

**Comment by albar965** [comment]
No problem. The internal times are already millisecond values with some exceptions where only seconds were stored. And this was the aircraft trail.
I intended to replace the old 32-bit seconds since Epoch variable anyway (there was a TODO remark).
Also found some issues with timezone conversions.

Just checked and time is now accurate across all of LNM aircraft trail. Tested with X-Plane, FSX and MSFS.
